Convereted lettuce tests into bokchoy for LTI Xblock

zrana opened the pull request.

zrana commented:
@estute @cpennington @jmbowman 
I am trying to convert LTI xblock lettuce test into bokchoy. Currently LTI lettuce tests are using a stub service to mock LTI provider, every time we execute a test it starts the LTI stub service on given host address(which is using django setting to getting host) and creates a post request.
I am trying to replicate the same in bokchoy by starting this stub in bok-choy settings file along with other stubs, but I get error messages on stub server response. If I try to execute LTI bokchoy test, I am getting this error message:
`Traceback (most recent call last):
  File "/usr/lib/python2.7/SocketServer.py", line 596, in process_request_thread
    self.finish_request(request, client_address)
  File "/usr/lib/python2.7/SocketServer.py", line 331, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib/python2.7/SocketServer.py", line 652, in __init__
    self.handle()
  File "/usr/lib/python2.7/BaseHTTPServer.py", line 340, in handle
    self.handle_one_request()
  File "/usr/lib/python2.7/BaseHTTPServer.py", line 328, in handle_one_request
    method()
  File "stubs/lti.py", line 65, in do_POST
    if self._check_oauth_signature(params, self.post_dict.get('oauth_signature', "")):
  File "stubs/lti.py", line 295, in _check_oauth_signature
    host = getattr(settings, 'LETTUCE_HOST', '127.0.0.1')
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/conf/__init__.py", line 56, in __getattr__
    self._setup(name)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/conf/__init__.py", line 39, in _setup
    % (desc, ENVIRONMENT_VARIABLE))
ImproperlyConfigured: Requested setting LETTUCE_HOST, but settings are not configured. You must either define the environment variable DJANGO_SETTINGS_MODULE or call settings.configure() before accessing settings.`

But if I hardcode the host address with 'edx.devstack.lms' in "edx-platform/common/djangoapps/terrain/stubs/lti.py" file and then execute the bokchoy test, I got "Verify HMAC-SHA1 failed: sig base string" error message in stub server response.
There must be some sort of setting which is being used by lettuce test but is not called when bok-choy runs the stub. I will be very grateful if you could provide some guidance on how to start the LTI stub properly in bok-choy tests.

zrana commented:
Thank you so much for the help. I will implement your suggestions and share the results with you. :+1:

zrana commented:
@jmbowman 

Thank you for your suggestions and taking time on this. I had added the suggested changes(oAuth settings) in bokchoy file, but still getting the same issue. I tried with different workarounds and also took help from local teams here, but nothing works. I have pushed the suggested changes and I would be very grateful if you could further take look into this to see what's missing. I know you are quite busy these days, but whenever you have some free time, please take a look.

estute commented:
@zrana I'm not sure if you are only checking locally, but the Jenkins bokchoy tests are failing on running migrations, before we even get to the point you mentioned. Are you aware of this?

edx-travis commented:
@zrana just following up here. Are you still running into issues with setting these up? Do you need any help? I checked the bok choy run and it looks like your tests are running now, just not passing.

estute commented:
@zrana just following up here. Are you still running into issues with setting these up? Do you need any help? I checked the bok choy run and it looks like your tests are running now, just not passing.

zrana commented:
@estute 
Oh, apologies. Sorry for bothering, I had reverted the settings in bokchoy file before pushing and assumed that it is working fine on jenkins. Now I have pushed new changes and migrations are running fine, but bokchoy test is still failing. I have added the changes in this [file](https://github.com/edx/edx-platform/pull/19429/files#diff-af1610807e9adf5015f264fc5f624934R301) also added the third party auth flag [here](https://github.com/edx/edx-platform/pull/19429/files#diff-50af9c04afe908e5df23c38d3ad994d0R208) as suggested by @bowman.

You can view the failing test [here](https://build.testeng.edx.org/job/edx-platform-bokchoy-pipeline-pr/186/testReport/junit/acceptance.tests.lms.test_lms_lti/TestLTIConusmer/Run_tests___22___test_lti_is_rendered_iframe/).

If you see the bok_choy_lti.log file [here](https://build.testeng.edx.org/job/edx-platform-bokchoy-pipeline-pr/186/artifact/test_root/log/shard_22/bok_choy_lti.log), We are still facing the same oAuth(HMAC-SHA1 failed) issue in LTI server response.

I would appreciate any help you can provide.

jmbowman commented:
jenkins run bokchoy
jenkins run quality

jmbowman commented:
I've reproduce the one bok-choy test failure in my local devstack, still debugging.  The quality failure can be ignored, that file is only used in tests and should be going away soon anyway.

jmbowman commented:
I think I found the problem - in the OAuth request, the value of `context_title` is set to `Test Course'';!--"<XSS>=&{()}247447616136770514322140762216933525886`, whereas when validating the signature it's been truncated to `Test Course\'\';!--`.  The lost portion is part of the XSS injection attempt from [here](https://github.com/edx/edx-platform/blob/7172bf46c0cfe0e25b66830568d732397a8e2348/common/test/acceptance/tests/helpers.py#L804).  Apparently something in the request handling process recognizes this as a potential problem and strips it out, but that means the OAuth signatures don't match.  I suspect it will work correctly if you use a non-problematic display name for the course fixture in the new test class.

If you need to insert some debugging code to verify the base strings and signatures being compared, the request preparation is done [here in the LTI consumer XBlock](https://github.com/edx/xblock-lti-consumer/blob/38ad154c0b6a041d24ecf3620a6b5c8bbb51691d/lti_consumer/oauth.py#L54-L59), and the signature verification is performed [here in the LTI stub server](https://github.com/edx/edx-platform/blob/7172bf46c0cfe0e25b66830568d732397a8e2348/common/djangoapps/terrain/stubs/lti.py#L306).  I ended up editing my local copy of these files and `oauthlib` to write some debugging information to a file in order to track this down.

zrana commented:
Hi @jmbowman @estute , it's really good of you to take time out to debug and solve this, I was stuck on it for a long time and I really appreciate your help. Thanks a lot.

jmbowman commented:
@zrana I noticed that you haven't tried changing the course display name yet; do you still need help with this?

zrana commented:
Thank you @jmbowman for all the help. I have tried locally and it worked. I haven't pushed the change yet. I am just finalizing all the LTI tests (there are lot of tests for LTI ).

zrana commented:
jenkins run js

zrana commented:
@jmbowman @estute Thank you again for your continued support. The issues is now fixed and I have pushed most of the LTI tests. There is flakiness in one test. I am fixing it, meanwhile if any of you can take a quick look that would be very helpful.

zrana commented:
jenkins run bokchoy
jenkins run js

zrana commented:
jenkins run bokchoy

zrana commented:
jenkins run bokchoy

saleem-latif commented on common/djangoapps/terrain/stubs/lti.py:
```diff
@@ -72,7 +77,8 @@ def do_POST(self):
 
-                host = getattr(settings, 'LETTUCE_HOST', self.server.server_address[0])
+                host = os.environ.get('BOK_CHOY_HOSTNAME', self.server.server_address[0])
+                # host = getattr(settings, 'LETTUCE_HOST', self.server.server_address[0])
```
Make sure to remove this when you are ready for merge.

saleem-latif commented on common/test/acceptance/pages/lms/courseware.py:
```diff
@@ -375,6 +377,82 @@ def word_cloud_answer_list(self):
+        )
+        return self.q(css='.action-preview-select')
+
+    def see_elem_text(self, elem_css):
```
nit: `get_elem_text `

saleem-latif commented on common/test/acceptance/pages/lms/courseware.py:
```diff
@@ -375,6 +377,82 @@ def word_cloud_answer_list(self):
+    @property
+    def switch_role_selector(self):
+        """
+        return role selector
```
nit: `get_role_selector` or something similar?

saleem-latif replied:
Change doc string to "Find and return role selector."

saleem-latif commented on common/test/acceptance/pages/lms/courseware.py:
```diff
@@ -375,6 +377,82 @@ def word_cloud_answer_list(self):
+
+    def see_elem_text(self, elem_css):
+        """
+        return text of element
```
nit: "return text of element" -> "Find the element with CSS selector given in the argument and return its text."

saleem-latif commented on common/test/acceptance/pages/lms/courseware.py:
```diff
@@ -375,6 +377,82 @@ def word_cloud_answer_list(self):
+        )
+        return self.q(css=elem_css).text[0]
+
+    def check_lti_component_no_elem(self, elem_css):
```
can you rename the method to something more expressive? like `is_lti_component_present`. Also, update the doc string according to the functionality.

saleem-latif commented on common/test/acceptance/pages/lms/courseware.py:
```diff
@@ -375,6 +377,82 @@ def word_cloud_answer_list(self):
+    def submit_lti_answer(self, button_css):
+        """ Click on Submit button"""
+        wait = WebDriverWait(self.browser, 10)
+        element = wait.until(EC.element_to_be_clickable((By.ID, button_css)))
```
See if you can use [this](https://github.com/edx/bok-choy/blob/951c546f85f89e4588aec9b051d7d0304a880213/bok_choy/page_object.py#L579) instead.

saleem-latif commented on common/test/acceptance/pages/lms/courseware.py:
```diff
@@ -375,6 +377,82 @@ def word_cloud_answer_list(self):
+        return content
+
+    @property
+    def check_role(self):
```
nit: `get_role`

saleem-latif commented on common/test/acceptance/pages/lms/courseware.py:
```diff
@@ -375,6 +377,82 @@ def word_cloud_answer_list(self):
+        """
+        return role text
+        """
+        role = self.q(css="h5").text[0]
```
nit: change this to
```
return self.q(css="h5").text[0]
```

saleem-latif commented on common/test/acceptance/pages/lms/instructor_dashboard.py:
```diff
@@ -1304,6 +1304,26 @@ def check_func():
+        """
+        return self.q(css='.grade-table').present
+
+    def see_value_in_the_grade_book(self, title, index):
```
nit: `see_value_in_the_grade_book` -> `get_value_in_the_grade_book `

zrana commented:
jenkins run bokchoy

zrana commented:
jenkins run bokchoy

zrana commented:
jenkins run bokchoy

zrana commented:
jenkins run bokchoy
jenkins run lettuce
jenkins run python
jenkins run quality

zrana commented:
jenkins run js

jmbowman commented:
🎉 Thanks for getting this done!  Is there going to be a separate PR for removing the original lettuce tests these were converted from?
